src/cargo/core/compiler/build_plan.rs

```diff
@@ -0,0 +1,136 @@
+//! A graph-like structure used to represent the rustc commands to build the project and the
+//! interdependencies between them.
+//!
+//! The BuildPlan structure is used to store the dependency graph of a dry run so that it can be
+//! shared with an external build system. Each Module in the BuildPlan comprises a single
+//! subprocess and defines the build environment, the outputs produced by the subprocess, and the
+//! dependencies on other Modules.
+
+use std::collections::BTreeMap;
+
+use super::{Context, Unit};
+use super::context::OutputFile;
+use util::{internal, CargoResult, ProcessBuilder};
+use std::sync::Arc;
+use std::path::PathBuf;
+use serde_json;
+
+#[derive(Debug, Serialize)]
+struct Module {
+    display_name: String,
+    deps: Vec<usize>,
+    outputs: Vec<PathBuf>,
+    links: BTreeMap<PathBuf, PathBuf>,
+    program: String,
+    args: Vec<String>,
+    env: BTreeMap<String, String>,
+    cwd: PathBuf,
+}
+
+#[derive(Debug, Serialize)]
+pub struct BuildPlan {
+    module_map: BTreeMap<String, usize>,
+    modules: BTreeMap<usize, Module>,
```

Comment:
I think we can make this `modules: Vec<Module>`, and then, in `get_id`, we could push a `Module::default` to this `Vec`. This will require `derive(Default)` as well.